### PR TITLE
batch transaction

### PR DESCRIPTION
**Pull request by diepmv**
for #5

### CODE REVIEW

**Comment by diepmv** [comment]
@Zergity help me review my test version

File: core/vm/evm.go [hunk 1]
```diff
@@ -177,89 +181,150 @@ func (evm *EVM) Interpreter() *EVMInterpreter {
...
+type Tx struct {
+	To    common.Address `json:"to"`
+	Data  []byte         `json:"data"`
+	Value *big.Int       `json:"value"`
+}
+
+var (
+	// CallBatchSignature is Keccak("callBatch([]Tx")
+	CallBatchSignature = []byte{0x7d, 0x15, 0x53, 0x2f}
+	BADDR              = common.HexToAddress("0x5555555555555555555555555555555555555555")
```
Comment (Zergity) [inline core/vm/evm.go:193]: this address is already declare somewhere else, please search for it

File: core/vm/evm.go [hunk 2]
```diff
@@ -177,89 +181,150 @@ func (evm *EVM) Interpreter() *EVMInterpreter {
...
+}
+
 // Call executes the contract associated with the addr with the given input as
 // parameters. It also handles any necessary value transfer required and takes
 // the necessary steps to create accounts and reverses the state in case of an
 // execution error or failed value transfer.
 func (evm *EVM) Call(caller ContractRef, addr common.Address, input []byte, gas uint64, value *big.Int) (ret []byte, leftOverGas uint64, err error) {
+	var txs []*Tx
+	if isBatchTx(addr, input) {
+		txs, _ = decodeBatchTxInput(input)
```
Comment (Zergity) [inline core/vm/evm.go:234]: don't ignore error here, this part should be more defensive: if the decode is not success, fallback to legacy tx

**Review by Zergity — CHANGES_REQUESTED** [review]
this logic should be applied in `func (st *StateTransition) TransitionDb()`, and only this specific line is need to called for each inner tx:

`ret, st.gas, vmerr = st.evm.Call(sender, st.to(), st.data, st.gas, st.value)`

**Comment by Zergity** [comment]
> `ret, st.gas, vmerr = st.evm.Call(sender, st.to(), st.data, st.gas, st.value)`

```
if contractCreation {
	ret, _, st.gas, vmerr = st.evm.Create(sender, st.data, st.gas, st.value)
} else {
	// Increment the nonce for the next transaction
	st.state.SetNonce(msg.From(), st.state.GetNonce(sender.Address())+1)
	if isBatch {
		var rets [][]byte
		for each txs {
			ret, st.gas, vmerr = st.evm.Call(sender, st.to(), st.data, st.gas, st.value)
			if vmerr != nil {
				break	// stop on the first error
			}
			rets.push(ret)
		}
		if vmerr == nil {
			ret = encodeRetsHere(rets)
		}
	} else {
		ret, st.gas, vmerr = st.evm.Call(sender, st.to(), st.data, st.gas, st.value)
	}
}
```

**Comment by diepmv** [comment]
@Zergity @enixdark help me review

File: core/state_transition.go [hunk 3]
```diff
@@ -325,12 +391,35 @@ func (st *StateTransition) TransitionDb() (*ExecutionResult, error) {
...
+
 	if contractCreation {
 		ret, _, st.gas, vmerr = st.evm.Create(sender, st.data, st.gas, st.value)
 	} else {
 		// Increment the nonce for the next transaction
 		st.state.SetNonce(msg.From(), st.state.GetNonce(sender.Address())+1)
-		ret, st.gas, vmerr = st.evm.Call(sender, st.to(), st.data, st.gas, st.value)
+
+		if isBatchTx(st.to(), st.data) {
+			snapshot := st.evm.StateDB.Snapshot()
```
Comment (Zergity) [inline core/state_transition.go:402]: only create snapshot after `var rets [][]byte`

File: core/state_transition.go [hunk 4]
```diff
@@ -325,12 +391,35 @@ func (st *StateTransition) TransitionDb() (*ExecutionResult, error) {
...
 	)
+
 	if contractCreation {
 		ret, _, st.gas, vmerr = st.evm.Create(sender, st.data, st.gas, st.value)
 	} else {
 		// Increment the nonce for the next transaction
 		st.state.SetNonce(msg.From(), st.state.GetNonce(sender.Address())+1)
-		ret, st.gas, vmerr = st.evm.Call(sender, st.to(), st.data, st.gas, st.value)
+
+		if isBatchTx(st.to(), st.data) {
```
Comment (Zergity) [inline core/state_transition.go:401]: `isBatchTx` and `decodeBatchTxInput` should be merged into one (`decodeBatchTx`) to simplify the if-else flow here

File: core/state_transition.go [hunk 5]
```diff
@@ -325,12 +391,35 @@ func (st *StateTransition) TransitionDb() (*ExecutionResult, error) {
...
+
+				for _, tx := range txs {
+					ret, st.gas, vmerr = st.evm.Call(sender, tx.To, tx.Data, st.gas, tx.Value)
+					if vmerr != nil {
+						st.evm.StateDB.RevertToSnapshot(snapshot)
+						break
+					}
+					rets = append(rets, ret)
+				}
+				ret, _ = encodeBatchTxOutput(rets)
```
Comment (Zergity) [inline core/state_transition.go:418]: as my comment for this, we only `encodeBatchTxOutput` when `vmerr == nil`. The test case is: send a batch tx of 3 tx, with the second tx failed, the result of the batch tx must be the revert reason of the failed tx.

File: core/state_transition.go [hunk 6]
```diff
@@ -266,6 +270,68 @@ func (st *StateTransition) preCheck() error {
...
+
+func isBatchTx(addr common.Address, input []byte) bool {
+	if input == nil || len(input) < 4 {
+		return false
+	}
+	return bytes.Equal(input[:4], callBatchSignature) && addr == params.EVMPP
+}
+
+func decodeBatchTxInput(encodedData []byte) (txs []*Tx, err error) {
+	batchAbi, err := abi.JSON(strings.NewReader(batchTxAbiJson))
```
Comment (Zergity) [inline core/state_transition.go:294]: please don't parse the JSON from text each time this method is called, create an `abi.ABI` struct contact for this

**Review by Zergity — CHANGES_REQUESTED** [review]

File: core/state_transition.go (hunk 3 above)
Comment (diepmv) [reply core/state_transition.go:411]: fixed

File: core/state_transition.go (hunk 4 above)
Comment (diepmv) [reply core/state_transition.go]: fixed

File: core/state_transition.go (hunk 5 above)
Comment (diepmv) [reply core/state_transition.go:423]: fixed

File: core/state_transition.go (hunk 6 above)
Comment (diepmv) [reply core/state_transition.go]: fixed

File: core/state_transition.go [hunk 7]
```diff
@@ -266,6 +269,85 @@ func (st *StateTransition) preCheck() error {
...
+	if err != nil {
+		return batchTx, txs, err
+	}
+	if err := json.Unmarshal(byteData, &txs); err != nil {
+		return batchTx, txs, err
+	}
+	return batchTx, txs, nil
+}
+
+func encodeBatchTxOutput(data [][]byte) ([]byte, error) {
```
Comment (Zergity) [inline core/state_transition.go:334]: I've added a commit in feature-tests PR, to test a high-level call and return value encode, please make sure it work with any return values.

[766925a475395064f1f9d2d9f82b6f89e1e124a3](https://github.com/ROIchain/feature-tests/pull/2/commits/766925a475395064f1f9d2d9f82b6f89e1e124a3)

**Review by Zergity — CHANGES_REQUESTED** [review]

File: core/state_transition.go [hunk 8]
```diff
@@ -266,6 +270,76 @@ func (st *StateTransition) preCheck() error {
...
+)
+
+func isBatchTx(addr common.Address, input []byte) bool {
+	if input == nil || len(input) < 4 {
+		return false
+	}
+	return bytes.Equal(input[:4], callBatchSignature) && addr == params.EVMPP
+}
+
+func decodeBatchTx(addr common.Address, encodedData []byte) (batchTx bool, txs []*Tx, err error) {
```
[the remark this answers is not in the document]
Comment (diepmv) [reply core/state_transition.go]: fixed

**Review by Zergity — CHANGES_REQUESTED** [review]
some return type is tightly packed by solidity, which is unptedictable from the evm view. please modify the code to return only the last tx result, instead of encoding all of them

**Review by Zergity — APPROVED** [review]
please add the gasLimit tests for features test and it's good to go

**Pull request merged**